shared/new_sort/MAPL_Sort.F90

```diff
@@ -0,0 +1,468 @@
+!       subroutine MAPL_Sort(A[Ci,Cl,Cr,Cd,DIM])
+!         integer(kind=[4,8]),           intent(INOUT) :: A (:,:)
+!         integer(kind=4)    , optional, intent(INOUT) :: Ci(:,:)
+!         integer(kind=8)    , optional, intent(INOUT) :: Cl(:,:)
+!         real   (kind=4)    , optional, intent(INOUT) :: Cr(:,:)
+!         real   (kind=8)    , optional, intent(INOUT) :: Cd(:,:)
+!         integer(kind=4),     optional, intent(IN   ) :: DIM
+!
+!\ev
+!   {\tt MAPL\_Sort} sorts the key (contained in a row or column of A)
+!   in ascending order and reorders the data in B or in non-key rows or columns of A
+!   in the same order; i.e., it does the same exchanges as were done 
+!   to the key in sorting it.  If, for example, on input B(:) contains the ordered integers
+!   from 1 to size(A), on output it will contain the positions of the elements of
+!   the sorted A in the unsorted A. In the last two signatures, DIM is the dimension
+!   of A or B being reordered. In the last signature, for example, DIM=1 corresponds
+!   to a B ordered as B(size(A),:), whereas DIM=2 corresponds to B(:,size(A)).
+!   The default is DIM=2. The quicksort is coded in C and does not appear here.
+
+!EOP
+!=============================================================================
+
+interface MAPL_Sort
+   module procedure SORTS
+   module procedure SORTSL
+   module procedure SORTSR
+   module procedure SORTSD
+   module procedure SORTL
+   module procedure SORTLL
+   module procedure SORTLR
+   module procedure SORTLD
+   module procedure SORT2AS
+   module procedure SORT2AL
+
+   module procedure SORT2LS
+   module procedure SORT2LL
+   module procedure SORT2LR
+   module procedure SORT2LD
+   module procedure SORT2SS
+   module procedure SORT2SR
+   module procedure SORT2SL
+   module procedure SORT2SD
+end interface
+
+contains
+
+subroutine SORTL(A,Bi,Bl,Br,Bd,Ci,Cl,Cr,Cd,Di,Dl,Dr,Dd,DIM)
+  integer(kind=8),           intent(INOUT) :: A(:)
+  integer(kind=4), optional, intent(INOUT) :: Bi(:)
+  integer(kind=8), optional, intent(INOUT) :: Bl(:)
+  real   (kind=4), optional, intent(INOUT) :: Br(:)
+  real   (kind=8), optional, intent(INOUT) :: Bd(:)
+  integer(kind=4), optional, intent(INOUT) :: Ci(:,:)
+  integer(kind=8), optional, intent(INOUT) :: Cl(:,:)
+  real   (kind=4), optional, intent(INOUT) :: Cr(:,:)
+  real   (kind=8), optional, intent(INOUT) :: Cd(:,:)
+  integer(kind=4), optional, intent(INOUT) :: Di(:,:)
+  integer(kind=8), optional, intent(INOUT) :: Dl(:,:)
+  real   (kind=4), optional, intent(INOUT) :: Dr(:,:)
+  real   (kind=8), optional, intent(INOUT) :: Dd(:,:)
+  integer,         optional, intent(IN   ) :: DIM
+
+  integer :: uDIM, sg, id, len
+  integer :: Dm(0,0)
+
+  len = size(A)
+
+  if    (present(Bi)) then
+     call QSORTL44(A,Bi,Dm,len,1,0)
+  elseif(present(Bl))then
+     call QSORTL84(A,Bl,Dm,len,1,0)
+  elseif(present(Br))then
+     call QSORTL44(A,Br,Dm,len,1,0)
+  elseif(present(Bd)) then
+     call QSORTL84(A,Bd,Dm,len,1,0)
```

Comment:
Yeah - no, this is a terrible/unacceptable style.    In theory all 4 of these args could be present.   The proper way todo this is to have separate interfaces for each case.